### ISSUE TITLE
Gist to convert python objects into pytorch tensors

### ISSUE THREAD

**Issue by sudarshan85** [GRAAL-Research/pytoune#13]
Hi,

I added a [gist](https://gist.github.com/sudarshan85/9647e375b7ba7af6313f0eb97ecb28fc) with a couple of functions to convert a numpy array (or python object such as list) into a pytorch tensor. I couldn't find anything like this in the library except a function going the other way around `pytoune.torch_to_numpy`. 

These functions are copied from the FastAI library ([core](https://github.com/fastai/fastai/blob/master/fastai/core.py) module) and modified a bit. If this is useful, I could submit a PR.

**Comment by freud14** [comment]
Hi,

Thank you for your suggestion. There is actually a converse for `pytoune.torch_to_numpy` called `numpy_to_torch`. See https://github.com/GRAAL-Research/pytoune/blob/master/pytoune/utils.py. I must have forgot to put it in the doc. By the way, PyToune calls it on the input in all functions so you don't need to call it by hand. 

Also, you will maybe notice that there is a `torch_to` function in that file. This function is also called on the input in all functions if you called `Model.to` beforehand. It is used to send to a GPU device all input tensors.

Let me know what you think.

Thank you.

Frédérik

**Comment by sudarshan85** [comment]
Ah, yes it doesn't seem to be in the docs. Thank you for that. 

> Also, you will maybe notice that there is a torch_to function in that file. This function is also called on the input in all functions if you called Model.to beforehand. It is used to send to a GPU device all input tensors.

So lets say my dataset consists of my training and validation tensors (that I created without calling `.cuda()`). Similarly, I created my pytorch module but haven't called `.cuda()`. In this scenario what all would I need to do to get everything to work on the GPU?

Currently, I manually create the tensors with `.cuda()` and I created my pytorch module with `.cuda()` (even though the Model class as a `cuda()` method, which is just a wrapper for pytorch's `.cuda()` method). I then create my model, dataloaders etc and call fit. This makes it run on the GPU. Does pytoune provide inbuilt methods for the tensors to be loaded on the GPU?

Also, just a note the name "Model" for the `Mode`l class is confusing with the pytorch model that I create. Because I find that model refers to the pytorch model created with `nn.Module`. In my code,I use the word `container` for my Model class object which seems to make more sense to me.

**Comment by freud14** [comment]
> So lets say my dataset consists of my training and validation tensors (that I created without calling `.cuda()`). Similarly, I created my pytorch module but haven't called `.cuda()`. In this scenario what all would I need to do to get everything to work on the GPU?
> ...
> Does pytoune provide inbuilt methods for the tensors to be loaded on the GPU?

Calling `.cuda()` or `.to()` on the `Model`class will result in both the PyTorch module and the training and validation tensors to be send to the GPU. 

>Also, just a note the name "Model" for the Modell class is confusing with the pytorch model that I create. Because I find that model refers to the pytorch model created with nn.Module. In my code,I use the word container for my Model class object which seems to make more sense to me.

I agree that it can be confusing. In my case, I call my PyTorch module `module` and my PyToune Model `model`.

**Comment by sudarshan85** [comment]
Ok thanks. I was able to figure that out. The problem was I was just calling `.cuda()` on the `Model` class and *not* saving it back. According to the code, the `.cuda()` method returns `self` which means I have to save it back to my model class which updates everything. 

Another issue I'm having (this might be a pytorch issue) is that I have to initialize some tensors to zero inside my pytorch module. When `.cuda()` is called on the `Model` class it does not transfer the internally initialized to the GPU (unless I explicitly call `.cuda()` on the tensors directly).  Is there a solution to this?

According to the docs the `numpy_to_torch` accepts any python object including list, tuple, and dict and returns a torch tensor. However, when I pass a list, I just get the list back instead of a tensor. But when I pass a numpy array, I get a tensor back. Is this how its supposed to work?

```
x = [1,2,3]
numpy_to_torch(x)
> [1,2,3]
x = np.array([1,2,3])
numpy_to_torch(x)
> tensor([1, 2, 3])

```

Finally, are there any plans for including `tqdm` progress tracker during training?

Thanks for your replies!

**Comment by freud14** [comment]
> Ok thanks. I was able to figure that out. The problem was I was just calling `.cuda()` on the `Model` class and *not* saving it back. According to the code, the `.cuda()` method returns `self` which means I have to save it back to my model class which updates everything. 

No, it is supposed to do it in-place. You don't need to "save" it.

> Another issue I'm having (this might be a pytorch issue) is that I have to initialize some tensors to zero inside my pytorch module. When `.cuda()` is called on the `Model` class it does not transfer the internally initialized to the GPU (unless I explicitly call `.cuda()` on the tensors directly).  Is there a solution to this?

You have to use `torch.zeros_like` or to use the `new_zeros` method on one of your tensors.

> According to the docs the `numpy_to_torch` accepts any python object including list, tuple, and dict and returns a torch tensor. However, when I pass a list, I just get the list back instead of a tensor. But when I pass a numpy array, I get a tensor back. Is this how its supposed to work?

The function supports Numpy arrays inside lists/tuples/dicts.

```
>>> x = [np.array([1,2,3]), np.array([4,5,6])]
>>> numpy_to_torch(x)
[tensor([1, 2, 3]), tensor([4, 5, 6])]
```

**Comment by sudarshan85** [comment]
> No, it is supposed to do it in-place. You don't need to "save" it. if you have an example where it does not do that, can you provide it?

Thanks. It works. For some reason it didn't work earlier.

> You have to use torch.zeros_like or to use the new_zeros method on one of your tensors.

Thanks for this! I was struggling how to fix this. `new_zeros` returned a tensor on the same device.

> The function supports Numpy arrays inside lists/tuples/dicts.

Ah. I misunderstood. Thank you. But I would think that adding functionality to convert direct python objects of numbers to tensors shouldn't be too hard. Would you be interested in PR for that?

**Comment by freud14** [comment]
> But I would think that adding functionality to convert direct python objects of numbers to tensors shouldn't be too hard. Would you be interested in PR for that?

The thing is that you would have to infer whether a list is just a list of objects and tensors of different sizes or it is list to convert into a tensor. I don't see how you could do it efficiently.

**Comment by sudarshan85** [comment]
After thinking about it, I agree with you. Besides I think its just natural for people to immediately deal with numpy arrays. Thanks for your help. This is a great helper library that I will be using a lot. This issue can be closed but just two quick questions.

1. Do you have an opinion on tqdm progress tracker for training? 
2. As far as I can tell there no visualization tools other than logger for tensorboard. Would it be interesting to have a plotting class of sorts with basic plotting like training/test loss and metrics etc?

**Comment by freud14** [comment]
> 1. Do you have an opinion on tqdm progress tracker for training? 

Oups sorry, I forgot to answer this question in one of your previous comments. It was actually in the plans to do it but I never worked on it. So if you would like to do a PR for that, it would be nice. Depending on if you can still display everything in the command line, I would integrate tqdm inside the ProgressCallback.

>2. As far as I can tell there no visualization tools other than logger for tensorboard. Would it be interesting to have a plotting class of sorts with basic plotting like training/test loss and metrics etc?

There's actually https://github.com/stared/livelossplot which provides a callback PyToune to display live loss plot in an IPython Notebook. It probably can work outside of a Notebook with some tweaks probably.
